### PR TITLE
udev: make OPTIONS="string_escape=replace" take effect on ENV{key}= assiginment

### LINKED ISSUE #19309: udev: slash in nvme model attribute leads to directories in /dev/disk/by-id
**systemd version the issue has been seen with**
 
Not actually sure, but afaict basically any version since by-id links for nvme devices was added in a5110c90303cf455db5062faef34d5724d12e2e9 up until git master will do this.

**Used distribution**

Ubuntu 20.04.

**Linux kernel version used** (`uname -a`)

5.4.0

**CPU architecture issue was seen on**
 
amd64

**Expected behaviour you didn't see**

Symlinks directly in /dev/disk/by-id

**Unexpected behaviour you saw**

As seen in https://discourse.maas.io/t/unable-to-deploy-valueerror-failed-to-find-storage-volume-id-nvme0n1/4442, a / in the model reported by a nvme device leads to a directory in /dev/disk/by-id.

**Steps to reproduce the problem**

 1. Find a NVMe device with a / in it's reported model name (this is the hard part)
 2. boot a system with it in
 3. find /dev/disk/by-id -type d

Maybe this is something that we need to cope with but it was certainly unexpected for me. I also can't see how to fix this by changing the udev rule, all I can think of is some variant of the "$attr{file}" syntax that indicates that escaping should be done, or that the kernel be changed to escape this value.

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #19309.

### CODE REVIEW

**Comment by mwhudson** [comment]
Hi, this does seem to fix the problem, thanks! I wasn't able to figure out how to make qemu set the model to something with a slash in, but it's easy to put / into the serial which produced the same issue, which was resolved by this patch.

File: man/udev.xml [hunk 1]
```diff
@@ -605,7 +605,7 @@
                   </listitem>
                 </varlistentry>
                 <varlistentry>
-                  <term><option>string_escape=<replaceable>none|replace</replaceable></option></term>
+                  <term><option>string_escape=<replaceable>none|replace|slash</replaceable></option></term>
```
Comment (poettering) [inline man/udev.xml]: this should explain what this does.

File: src/udev/udev-rules.c [hunk 2]
```diff
@@ -2012,6 +2018,19 @@ static int udev_rule_apply_token_to_event(
...
+                                if (q[0] == '\\')
+                                        q += q[1] != '\0';
+                                else if (q[0] == '/') {
+                                        q[0] = '_';
+                                        count++;
+                                }
+                        if (count > 0)
+                                log_rule_debug(dev, rules, "Replaced %zu shash(es) from result of ENV{%s}%s=\"%s\"",
+                                               count, name, token->op == OP_ADD ? "+" : "", token->value);
+                }
```
Comment (poettering) [inline src/udev/udev-rules.c:2032]: maybe make this a function of its own?

**Comment by poettering** [comment]
Hmm, so the other ID_SERIAL values (i.e. for usb and such) all are careful in escaping all weird chars with udev_replace_chars()

Only for nvme we don't do this, because it's done in a udev rule instead of helper. I figure the right fix for this would be to use udev_replace_chars() here too, i.e. escape these serials the same way as all others.

This would probably mean we'd have to add this kind of escaping to rules, too, at leas optional.

/cc @msekletar who added the original NVME patch for this in a5110c90303cf455db5062faef34d5724d12e2e9

I wonder how much the mangling for this would affect the serial strings IRL. At least on my system the serials don#t have such weird chars, so I guess in many cases we'll just be fine.

**Comment by poettering** [comment]
long story short: i thin the approach is OK, but we should go further and process the strings with the full power of udev_replace_chars() instead of a lightweight escaper that only processes "/".

**Comment by yuwata** [comment]
@poettering Thank you for your review, and sorry for late response. Now, `string_escape=replace` option is reused for `ENV{}=` assignment. As I commented in the commit message, strictly speaking, this breaks backward compatibility. PTAL.

File: src/udev/udev-rules.c [hunk 3]
```diff
@@ -2022,6 +2022,12 @@ static int udev_rule_apply_token_to_event(
                         l = strpcpyl(&p, l, val, " ", NULL);
 
                 (void) udev_event_apply_format(event, token->value, p, l, false);
+                if (event->esc == ESCAPE_REPLACE) {
```
Comment (poettering) [inline src/udev/udev-rules.c:2027]: hmm, if this is done for ENV{} assignments too now, I fiugre the man page explanation for string_escape= should be slightly updated.

it currently says:

>  Usually, control and other possibly unsafe characters are replaced in strings used for device naming. The mode of replacement can be specified with this option.

ENV{} doesn't really fit under the "strings used for device naming" umbrella. 

also, it should probably say that for the device naming stuff we default to escaping but for the env stuff we default to not escaping.

**Comment by poettering** [comment]
hmm, thinking about this more: can this actually work? i.e. wouldn't the ENV{} assignment ecape once and then the SYMLINKS= stuff escape a second time? so we'd end up with a doubly-escaped symlink?

**Comment by yuwata** [comment]
Updated. Now a new netif naming scheme is also added. PTAL.

> hmm, thinking about this more: can this actually work? i.e. wouldn't the ENV{} assignment ecape once and then the SYMLINKS= stuff escape a second time? so we'd end up with a doubly-escaped symlink?

@poettering Why double escaping do you think causes problem here? `string_escape=replace` just replaces characters, not 'real' escaping, so the operation is idempotent.

File: src/udev/udev-rules.c [hunk 4]
```diff
@@ -2053,21 +2053,20 @@ static int udev_rule_apply_token_to_event(
                 if (token->op == OP_ASSIGN_FINAL)
                         event->name_final = true;
 
+                if (sd_device_get_ifindex(dev, NULL) < 0) {
+                        log_rule_error(dev, rules,
+                                       "Only network interface can be renamed, ignoring NAME=\"%s\"; please fix it.",
```
Comment (keszybz) [inline src/udev/udev-rules.c:2065]: "please fix it" is unclear: fix what and how? Also, the fact that this is printed in red is enough to tell the reader that it's something to fix.

> `"Only network interfaces can be renamed, ignoring NAME=\"%s\"."`

**Review by keszybz — APPROVED** [review]
LGTM.

The log message can always be adjusted after -rc2, the behaviour change is what is important.

**Pull request merged**